controllers/mysqlcluster_controller_test.go

```diff
@@ -76,44 +112,537 @@ spec:
+			isUpdated, err := reconciler.createOrUpdateRBAC(ctx, reconciler.Log, cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(isUpdated).Should(BeFalse())
+		})
+
+		It("should create service account", func() {
+			isUpdated, err := reconciler.createOrUpdateRBAC(ctx, reconciler.Log, cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(isUpdated).Should(BeTrue())
 
-			expectedPorts := []corev1.ServicePort{
+			sa := &corev1.ServiceAccount{}
+			err = k8sClient.Get(ctx, client.ObjectKey{Name: serviceAccountPrefix + moco.UniqueName(cluster), Namespace: cluster.Namespace}, sa)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			isUpdated, err = reconciler.createOrUpdateRBAC(ctx, reconciler.Log, cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(isUpdated).Should(BeFalse())
+		})
+	})
+
+	Context("Agent token", func() {
+		It("should create agent token", func() {
+			isUpdated, err := reconciler.generateAgentToken(ctx, reconciler.Log, cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(isUpdated).Should(BeTrue())
+
+			err = k8sClient.Get(ctx, client.ObjectKey{Name: clusterName, Namespace: namespace}, cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(cluster.Status.AgentToken).ShouldNot(BeEmpty())
+
+			isUpdated, err = reconciler.generateAgentToken(ctx, reconciler.Log, cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(isUpdated).Should(BeFalse())
+		})
+	})
+
+	Context("StatefulSet", func() {
+		It("should create statefulset", func() {
+			serverIDBase := mathrand.Uint32()
+			cluster.Status.ServerIDBase = &serverIDBase
+
+			isUpdated, err := reconciler.createOrUpdateStatefulSet(ctx, reconciler.Log, cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(isUpdated).Should(BeTrue())
+
+			sts := &appsv1.StatefulSet{}
+			err = k8sClient.Get(ctx, client.ObjectKey{Name: moco.UniqueName(cluster), Namespace: cluster.Namespace}, sts)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			var mysqldContainer *corev1.Container
+			var agentContainer *corev1.Container
+			for i, c := range sts.Spec.Template.Spec.Containers {
+				if c.Name == "mysqld" {
+					mysqldContainer = &sts.Spec.Template.Spec.Containers[i]
+				} else if c.Name == "agent" {
+					agentContainer = &sts.Spec.Template.Spec.Containers[i]
+				}
+			}
+			Expect(mysqldContainer).ShouldNot(BeNil())
+			Expect(agentContainer).ShouldNot(BeNil())
+
+			var claim *corev1.PersistentVolumeClaim
+			for i, v := range sts.Spec.VolumeClaimTemplates {
+				if v.Name == mysqlDataVolumeName {
+					claim = &sts.Spec.VolumeClaimTemplates[i]
+				}
+			}
+			Expect(claim).ShouldNot(BeNil())
+			Expect(claim.Spec.Resources.Requests.Storage().Value()).Should(BeNumerically("==", 1<<30))
+
+			isUpdated, err = reconciler.createOrUpdateStatefulSet(ctx, reconciler.Log, cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(isUpdated).Should(BeFalse())
+		})
+
+		It("should return error, when template does not contain mysqld container", func() {
```

Comment:
This constraint doesn't appear in the docs.  We should write it in `crd_mysqlcluster.md`.